solidity/contracts/bridge/VendingMachine.sol

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import "../token/TBTCToken.sol";
+import "../token/IReceiveApproval.sol";
+import "../GovernanceUtils.sol";
+
+/// @title TBTC v2 Vending Machine
+/// @notice The Vending Machine is the owner of TBTC v2 token and can mint
+///         TBTC v2 tokens in 1:1 ratio from TBTC v1 tokens. TBTC v2 can be
+///         unminted back to TBTC v1 with or without a fee - fee parameter is
+///         controlled by the Governance. This implementation acts as a bridge
+///         between TBTC v1 and TBTC v2 token, allowing to mint TBTC v2 before
+///         the system is ready and fully operational without sacrificing any
+///         security guarantees and decentralization of the project.
+///         Vending Machine as a contract itself is not upgradeable, though
+///         TBTC v2 token ownership can be updated in a two-step,
+///         governance-controlled process. It is expected that this process
+///         will be executed before the v2 system launch.
+contract VendingMachine is Ownable, IReceiveApproval {
```

Comment:
What will happen with all the TBTCv1 tokens locked in this contract? Do we want to have a possibility to forward them to the new `VendingMachine` contract or burn them?

Reply:
The new vending machine should receive them, it was supposed to be the step of the finalize update process but I missed it!

Reply:
This question escalated to two hours offline discussion with @nkuba about governance and possible attacks on the vending machine. The main problem is that updated vending machine can run any code (it can even be an EOA) and even though we do not transfer any tBTC v1 to the updated vending machine in the current version of the contract, compromised governance or updated vending machine that becomes an owner of TBTC v2 token can print TBTC v2 and use the printed TBTC v2 to unmint TBTC v1 from this vending machine version. Unmint fee update imposes a threat as well as it can be used to stop unmints until a malicious upgrade is finalized.

User funds safety should be treated with the highest priority and at the same time, we need to have a way to publicly iterate on tBTC v2 version. We came up with the following solution we believe is the best one:
- Vending machine upgrade: dev team initializes (proposes) a change, governance finalizes (approves).
- Fee update: governance initializes a change, governance finalizes the change, dev team has a veto right (only in the current vending machine contract).
- Governable delay will be increased to 7 days.
- TBTC v1 will be transferred to the updated vending machine version.

Reply:
After some additional consideration, we decided to not introduce the vetoer role for unmint fee update. The development team will initiate all changes based on the request from DAO and DAO will finalize the change after the governance delay.